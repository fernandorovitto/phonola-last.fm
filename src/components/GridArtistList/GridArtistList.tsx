import 'react-responsive-carousel/lib/styles/carousel.min.css';
import { GridList, GridListTile, GridListTileBar } from '@material-ui/core';
import React, { MouseEvent } from 'react';
import { RouteComponentProps, withRouter } from 'react-router';
import {
  searchArtistInfoAction,
  searchArtistTopAlbumsAction,
} from '@Store/artists/actions';
import { Artist } from '@Store/artists/types';
import defaultArtistCover from '@Assets/images/defaultArtistCover.png';
import { IBrowser } from 'redux-responsive/types';
import styled from 'styled-components';

interface PropsFromState extends RouteComponentProps {
  artists: Artist[];
  browser: IBrowser;
}

interface PropsFromDispatch {
  searchArtistTopAlbums: typeof searchArtistTopAlbumsAction;
  searchArtistInfo: typeof searchArtistInfoAction;
}

export type IProps = PropsFromState & PropsFromDispatch;

const gridWidthColumns: Record<string, number> = {
  extraSmall: 2,
  small: 3,
  // eslint-disable-next-line sort-keys
  medium: 4,
  // eslint-disable-next-line sort-keys
  large: 5,
  // eslint-disable-next-line sort-keys
  infinity: 6,
};

class GridArtistList extends React.Component<IProps> {
  private _clickArtist = (event: MouseEvent, artist: Artist): void => {
    event.preventDefault();
    this.props.searchArtistInfo(artist.mbid, artist.name);
    this.props.searchArtistTopAlbums(artist.mbid, artist.name);
    this.props.history.push({
      pathname: '/artist',
    });
  };

  private _getImage = (image: string): string =>
    image !== '' ? image : defaultArtistCover;

  private _gridItem = (artist: Artist, index: number): React.ReactElement => (
    <GridListTile
      key={index}
      cols={1}
      onClick={(e: MouseEvent): void => this._clickArtist(e, artist)}
    >
      <ImageContainer
        src={this._getImage(artist.avatar.xl)}
        alt={artist.name}
      />
      <StyledGridListTileBar
        title={artist.name}
        subtitle={
          <span>
            Listeners:{' '}
            {artist.listeners.toLocaleString(undefined, {
              maximumFractionDigits: 2,
            })}
          </span>
        }
        actionPosition="left"
      />
    </GridListTile>
  );

  private _gridItems = (): React.ReactNode =>
    this.props.artists.map(
      (artist: Artist, index: number): React.ReactElement =>
        this._gridItem(artist, index),
    );

  private _getGridColumns = (type: string): number => gridWidthColumns[type];

  public render(): React.ReactElement {
    return (
      <Wrapper>
        <StyledGridList
          spacing={1}
          cols={this._getGridColumns(this.props.browser.mediaType)}
        >
          {this._gridItems()}
        </StyledGridList>
      </Wrapper>
    );
  }
}

export default withRouter(GridArtistList);

// Styled components

const Wrapper = styled('div')`
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  padding: 0 20px;
`;

const StyledGridList = styled(GridList)``;

const StyledGridListTileBar = styled(GridListTileBar)`
  height: 80px;
`;

const ImageContainer = styled('img')`
  border: 1px rgba(255, 255, 255, 0.5) solid;
  max-height: 100%;
  max-width: 100%;
`;
