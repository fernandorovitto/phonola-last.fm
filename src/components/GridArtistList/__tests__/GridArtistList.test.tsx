import 'jest-styled-components';
import GridArtistList, { IProps } from '../GridArtistList';
import { Artist } from '@Store/artists/types';
import { artists } from '@Utils/test/models/artists';
import { browser } from '@Utils/test/models/browser';
import { createMemoryHistory } from 'history';
import { location } from '@Utils/test/models/location';
import { match } from '@Utils/test/models/match';
import React from 'react';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router } from 'react-router-dom';
describe('Test for GridArtistList Component', () => {
  const defaultArtists: Artist[] = [] as Artist[];
  const history = createMemoryHistory();

  const defaultProps: IProps = {
    artists: defaultArtists,
    browser,
    history,
    location,
    match,
    searchArtistInfo: jest.fn(),
    searchArtistTopAlbums: jest.fn(),
  };

  test('Should render the component with no artists', () => {
    const tree = renderer.create(
      <Router>
        <GridArtistList {...defaultProps} />
      </Router>,
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  test('Should render the component with artists', () => {
    const props: IProps = {
      ...defaultProps,
      artists,
    };
    const tree = renderer.create(
      <Router>
        <GridArtistList {...props} />
      </Router>,
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
