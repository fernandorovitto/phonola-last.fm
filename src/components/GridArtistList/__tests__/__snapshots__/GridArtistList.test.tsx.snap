// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for GridArtistList Component Should render the component with artists 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0 20px;
}

.c2 {
  height: 80px;
}

.c1 {
  border: 1px rgba(255,255,255,0.5) solid;
  max-height: 100%;
  max-width: 100%;
}

<div
  className="c0"
>
  <ul
    className="MuiGridList-root "
    style={
      Object {
        "margin": -0.5,
      }
    }
  >
    <li
      className="MuiGridListTile-root"
      onClick={[Function]}
      style={
        Object {
          "height": 181,
          "padding": 0.5,
          "width": "16.666666666666668%",
        }
      }
    >
      <div
        className="MuiGridListTile-tile"
      >
        <img
          alt="Megadeth"
          className="c1"
          src="test-file-stub"
        />
        <div
          className="MuiGridListTileBar-root c2 MuiGridListTileBar-titlePositionBottom MuiGridListTileBar-rootSubtitle"
        >
          <div
            className="MuiGridListTileBar-titleWrap"
          >
            <div
              className="MuiGridListTileBar-title"
            >
              Megadeth
            </div>
            <div
              className="MuiGridListTileBar-subtitle"
            >
              <span>
                Listeners:
                 
                10,000
              </span>
            </div>
          </div>
        </div>
      </div>
    </li>
    <li
      className="MuiGridListTile-root"
      onClick={[Function]}
      style={
        Object {
          "height": 181,
          "padding": 0.5,
          "width": "16.666666666666668%",
        }
      }
    >
      <div
        className="MuiGridListTile-tile"
      >
        <img
          alt="Pantera"
          className="c1"
          src="test-file-stub"
        />
        <div
          className="MuiGridListTileBar-root c2 MuiGridListTileBar-titlePositionBottom MuiGridListTileBar-rootSubtitle"
        >
          <div
            className="MuiGridListTileBar-titleWrap"
          >
            <div
              className="MuiGridListTileBar-title"
            >
              Pantera
            </div>
            <div
              className="MuiGridListTileBar-subtitle"
            >
              <span>
                Listeners:
                 
                10,000
              </span>
            </div>
          </div>
        </div>
      </div>
    </li>
  </ul>
</div>
`;

exports[`Test for GridArtistList Component Should render the component with no artists 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0 20px;
}

<div
  className="c0"
>
  <ul
    className="MuiGridList-root "
    style={
      Object {
        "margin": -0.5,
      }
    }
  />
</div>
`;
